### DISCUSSION TITLE
topological_sort: differences between versions 2.5 and 2.6

### DISCUSSION THREAD

**Discussion by tyralla** [networkx/networkx#4987] [Q&A] [answered]
The following code produces different results for the two mentioned NetworkX versions:

```python
digraph = networkx.DiGraph()
digraph.add_nodes_from(["C1", "C2", "B", "A"])
digraph.add_edge("A", "B")
digraph.add_edge("B", "C1")
digraph.add_edge("B", "C2")
print(list(networkx.topological_sort(digraph)))
```

For 2.5.1: `['A', 'B', 'C2', 'C1']`

For 2.6.2: `['A', 'B', 'C1', 'C2']`

Both results are correct, but I am wondering what is causing the difference.  Does the new and the future NetworkX versions make any promises on the sorted order of nodes that are "unsortable" based on the network structure alone?

**Comment by dschult** [comment, marked as answer]
The difference is a change (#4757) that
- allowed topological generations to be discovered while creating the topological sort, and
- handled multigraphs correctly.

One part of that change is that the order of each generation is reversed compared to what it used to be (we used to pop off the top of the list until empty and now we iterate over the list). Both answers are correct. No order of nodes is promised. The order depends on the topology of the graph, and the details of how we loop over the nodes at each generation of the process.

**Reply by tyralla** [reply]
Thanks for the explanation. The link was helpful. The `topological_generations` algorithm seems like a good alternative to me, as it allows me to sort the nodes of each generation by myself (based on their names or so).

**Reply by dschult** [reply]
That kind of flexibility is one good reason for `topological_generations` -- though I hadn't thought of sorting the generations.
